tokio-executor/src/thread_pool/park.rs

```diff
@@ -0,0 +1,181 @@
+use crate::loom::sync::atomic::AtomicUsize;
+use crate::loom::sync::atomic::Ordering::SeqCst;
+use crate::loom::sync::{Arc, Condvar, Mutex};
+use crate::park::{Park, Unpark};
+
+use std::error::Error;
+use std::fmt;
+use std::time::Duration;
+
+/// Parks the thread.
+#[derive(Debug)]
+pub(crate) struct DefaultPark {
+    inner: Arc<Inner>,
+}
+
+/// Unparks threads that were parked by `DefaultPark`.
+#[derive(Debug, Clone)]
+pub(crate) struct DefaultUnpark {
+    inner: Arc<Inner>,
+}
+
+/// Error returned by [`ParkThread`]
+///
+/// This currently is never returned, but might at some point in the future.
+///
+/// [`ParkThread`]: struct.ParkThread.html
+#[derive(Debug)]
+pub(crate) struct ParkError {
+    _p: (),
+}
+
+const EMPTY: usize = 0;
+const PARKED: usize = 1;
+const NOTIFIED: usize = 2;
+
+#[derive(Debug)]
+struct Inner {
+    state: AtomicUsize,
+    lock: Mutex<()>,
+    cvar: Condvar,
+}
+
+impl DefaultPark {
+    /// Creates a new `DefaultPark` instance.
+    pub(crate) fn new() -> DefaultPark {
+        DefaultPark {
+            inner: Arc::new(Inner {
+                state: AtomicUsize::new(EMPTY),
+                lock: Mutex::new(()),
+                cvar: Condvar::new(),
+            }),
+        }
+    }
+}
+
+impl Park for DefaultPark {
+    type Unpark = DefaultUnpark;
+    type Error = ParkError;
+
+    fn unpark(&self) -> Self::Unpark {
+        let inner = self.inner.clone();
+        DefaultUnpark { inner }
+    }
+
+    fn park(&mut self) -> Result<(), Self::Error> {
+        self.inner.park(None);
+        Ok(())
+    }
+
+    fn park_timeout(&mut self, duration: Duration) -> Result<(), Self::Error> {
+        self.inner.park(Some(duration));
+        Ok(())
+    }
+}
+
+impl Unpark for DefaultUnpark {
+    fn unpark(&self) {
+        self.inner.unpark();
+    }
+}
+
+impl fmt::Display for ParkError {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.description().fmt(fmt)
```

Comment:
```suggestion
        write!(fmt, "unknown park error")
```

nit: `Error::description` is deprecated (see #1245)